src/raftstore/store/store.rs

```diff
@@ -1838,49 +1853,148 @@ impl<T: Transport, C: PdClient> Store<T, C> {
-            info!(
-                "{} epoch changed {:?} != {:?}, need re-check later",
-                peer.tag,
-                region.get_region_epoch(),
-                epoch
-            );
-            return;
+        if let Some(epoch) = epoch {
+            if region.get_region_epoch().get_version() != epoch.get_version() {
+                error!(
+                    "[region {}] {} epoch changed {:?} != {:?}, need re-check later",
+                    region_id,
+                    peer.tag,
+                    region.get_region_epoch(),
+                    epoch
+                );
+                cb.map(|cb| {
+                    cb(new_error(box_err!(
+                        "[region {}] {} epoch changed {:?} != {:?}, need re-check later",
+                        region_id,
+                        peer.tag,
+                        region.get_region_epoch(),
+                        epoch
+                    )))
+                });
+                return;
+            }
         }
 
-        let key = keys::origin_key(&split_key);
-        let task = PdTask::AskSplit {
-            region: region.clone(),
-            split_key: key.to_vec(),
-            peer: peer.peer.clone(),
-            right_derive: self.cfg.right_derive_when_split,
-        };
+        match self.split_callbacks.entry(split_key.clone()) {
+            HashMapEntry::Occupied(_) => {
+                error!(
+                    "[region {}] is already splitting at {:?}",
+                    region_id,
+                    split_key
+                );
+                cb.map(|cb| {
+                    cb(new_error(box_err!("already splitting at {:?}", split_key)));
+                });
+                return;
+            }
+            HashMapEntry::Vacant(entry) => {
+                let task = PdTask::AskSplit {
+                    region: region.clone(),
+                    split_key: split_key.clone(),
+                    peer: peer.peer.clone(),
+                    right_derive: self.cfg.right_derive_when_split,
+                };
 
-        if let Err(e) = self.pd_worker.schedule(task) {
-            error!(
-                "{} failed to notify pd to split at {:?}: {}",
-                peer.tag,
-                split_key,
-                e
-            );
+                if let Err(e) = self.pd_worker.schedule(task) {
+                    error!(
+                        "[region {}] {} failed to notify pd to split at {:?}: {}",
+                        region_id,
+                        peer.tag,
+                        split_key,
+                        e
+                    );
+                    cb.map(|cb| {
+                        cb(new_error(
+                            box_err!("failed to split at {:?}: {}", split_key, e),
+                        ))
+                    });
+                } else {
+                    entry.insert(cb);
```

Comment:
we should pass the callback to the PD and let PD worker passes it as the split request callback to the raftstore.

Assume we have following cases:

1. AskSplit and insert to callback hash map.
2. PD workers sends a split request to raft store
3. The peer became follower, so the split request can be executed
4. The callback in the hash map can be called.